spec/function.dd

```diff
@@ -173,124 +173,317 @@ int foo();
+$(LI The returned type allows indirect access$(MDASH)through `ref` return and/or by transitively
+following class references, dynamic arrays, or pointers$(MDASH)to unqualified data or data with
+qualifiers other than `immutable`. Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+// These functions are weakly pure:
+pure Object f1(int);
+pure int* f2(int);
+pure const(int[]) f3(int);
+struct S
+{
+    const int x;
+}
+pure S* f4(int);
+---
+)
+
+$(LI The function is a constructor or destructor.)
+)
+
+$(H3 Strongly Pure Functions and Optimizations)
+
+$(P Strong purity enables optimizations based on their inability to read or modify data outside
+their parameters. In turn, that guarantees the function always returns the same result for the
+same arguments.)
+
+$(P A strongly pure function that accepts and returns by value can be assumed to return the same
+value for the same inputs. The compiler is allowed to memoize the result and reuse it for
+subsequent calls with identical arguments. Example:)
+
+---
+pure double fun(double);
+auto x = fun(1) + fun(1);
+// Code generated may be:
+// auto __t = fun(1);
+// auto x = __t + __t;
+---
+
+$(P For the purpose of deeming two calls to a strongly pure function equivalent, all arguments must
+be bitwise equal (as per the `is` operator), not logically equal (as per the `==` operator).
+Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+pure double fun(double x) { return x * 42; }
+pure int gun(int x) { return x ^ 42; }
+pure string hun(string s) { return s; }
+auto a = fun(1.0) + fun(1.0); // can memoize
+assert(-0.0 == 0.0 && -0.0 !is 0.0);
+auto b = fun(-0.0) + fun(0.0); // can NOT memoize, -0.0 and 0.0 have different bit patterns
+auto c = hun("ab") ~ hun("ab"); // can memoize ONLY if the two strings are interned
+---
+)
+
+$(P An implementation may not evaluate a strongly pure function speculatively. This is because a
+strongly pure function may abort the program, throw, or never terminate. Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+pure double fun(double x) { return x + 42; }
+pure double gun(size_t n)
+{
+    double result = 0;
+    foreach (i; 0 .. 100)
+    {
+        if (i == n) break;
+        result += fun(42);
+    }
+    return result;
+/*
+This code is NOT equivalent because it evaluates fun speculatively for n == 0:
+    double result = 0;
+    double __t = fun(42);
+    foreach (i; 0 .. 100)
+    {
+        if (i == n) break;
+        result += __t;
+    }
+    return result;
+*/
+}
+---
+)
+
+$(P A call to a `nothrow` strongly pure function that accepts and returns by value can be moved
+relative to its surrounding code, as long as its data dependencies are respected and there is no
+change in control flow. Example:)
+
+---
+pure nothrow double fun(double);
+pure nothrow double gun(double);
+void hun(double x)
+{
+    auto y = fun(x) + gun(x); // evaluate fun(x) and gun(x) in any order
+    foreach (i; 0 .. 100)
+    {
+        y += fun(10); // may hoist fun(10) outside the loop
```

Comment:
What if 'fun' here aborts or never terminates?
It's upsetting that pure functions are allowed to abort, that feels like an impure operation by definition... Never terminating seems worthless for a pure function, so I'm not sure what value considering that case could bring?

Reply:
What? I ruined my own past commit... will get back.

Reply:
Oh, the text is actually correct. I thought I'd messed up an earlier edit about speculative execution.

The call can be hoisted because it will be executed at least once anyway. I'll add clarifying text.

Reply:
@TurkeyMan added clarifying text